File: apps/desktop/src/lib/backend/github.ts
import { invoke } from '$lib/backend/ipc';

export type Verification = {
	user_code: string;
	device_code: string;
};

export async function initDeviceOauth() {
	return await invoke<Verification>('init_device_oauth');
}

export async function checkAuthStatus(params: { deviceCode: string }) {
	return await invoke<string>('check_auth_status', params);
}

File: crates/gitbutler-tauri/src/github.rs
use but_api::{
    NoParams,
    commands::github::{self},
    error::Error,
    github::{AuthStatusResponseSensitive, AuthenticatedUserSensitive, GetGhUserParams},
};
use but_github::{CheckAuthStatusParams, Verification};
use tracing::instrument;

#[tauri::command(async)]
#[instrument(err(Debug))]
pub async fn init_device_oauth() -> Result<Verification, Error> {
    github::init_device_oauth(NoParams {}).await
}

#[tauri::command(async)]
#[instrument(err(Debug))]
pub async fn check_auth_status(device_code: String) -> Result<AuthStatusResponseSensitive, Error> {
    github::check_auth_status(CheckAuthStatusParams { device_code }).await
}

#[tauri::command(async)]
#[instrument(err(Debug))]
pub async fn get_gh_user(username: String) -> Result<Option<AuthenticatedUserSensitive>, Error> {
    github::get_gh_user(GetGhUserParams { username }).await
}
